packages/utils/src/lib/merge-configs.ts

```diff
@@ -0,0 +1,229 @@
+      ...mergePlugins(acc.plugins, obj.plugins),
+      ...mergePersist(acc.persist, obj.persist),
+      ...mergeUpload(acc.upload, obj.upload),
+    }),
+    config,
+  );
+}
+
+function mergeCategories(
+  a: CategoryConfig[] | undefined,
+  b: CategoryConfig[] | undefined,
+) {
+  if (!a && !b) {
+    return {};
+  }
+
+  const mergedMap = new Map<string, CategoryConfig>();
+
+  const addToMap = (categories: CategoryConfig[]) => {
+    categories.forEach(newObject => {
+      if (mergedMap.has(newObject.slug)) {
+        const existingObject: CategoryConfig | undefined = mergedMap.get(
+          newObject.slug,
+        );
+
+        mergedMap.set(newObject.slug, {
+          ...existingObject,
+          ...newObject,
+
+          refs: mergeBySlugs(existingObject?.refs, newObject.refs),
+        });
+      } else {
+        mergedMap.set(newObject.slug, newObject);
+      }
+    });
+  };
+
+  if (a) {
+    addToMap(a);
+  }
+  if (b) {
+    addToMap(b);
+  }
+
+  // Convert the map back to an array
+  return { categories: [...mergedMap.values()] };
+}
+
+function mergePlugins(
+  a: PluginConfig[] | undefined,
+  b: PluginConfig[] | undefined,
+) {
+  if (!a && !b) {
+    return {};
+  }
+
+  const mergedMap = new Map<string, PluginConfig>();
+
+  const addToMap = (plugins: PluginConfig[]) => {
+    plugins.forEach(newObject => {
+      if (mergedMap.has(newObject.slug)) {
+        const existingObject: PluginConfig | undefined = mergedMap.get(
+          newObject.slug,
+        );
+
+        mergedMap.set(newObject.slug, {
+          ...existingObject,
+          ...newObject,
+          // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+          ...(newObject.audits && {
+            audits: mergeBySlugs(existingObject?.audits, newObject.audits),
+          }),
+          ...((existingObject?.groups ?? newObject.groups) && {
+            groups: mergeGroupsBySlugs(
+              existingObject?.groups,
+              newObject.groups,
+            ),
+          }),
+        });
+      } else {
+        mergedMap.set(newObject.slug, newObject);
+      }
+    });
+  };
+
+  if (a) {
+    addToMap(a);
+  }
+  if (b) {
+    addToMap(b);
+  }
+
+  return { plugins: [...mergedMap.values()] };
+}
+
+function mergeGroupRefsBySlugs<T extends GroupRef>(
+  a: T[] | undefined,
+  b: T[] | undefined,
+) {
+  const map = new Map<string, T>();
+
+  const addToMap = (groups: T[]) => {
+    groups.forEach(group => {
+      if (map.has(group.slug)) {
+        map.set(group.slug, { ...map.get(group.slug), ...group });
+      } else {
+        map.set(group.slug, group);
+      }
+    });
+  };
+
+  // Add objects from both arrays to the map
+  if (a) {
+    addToMap(a);
+  }
+  if (b) {
+    addToMap(b);
+  }
+
+  return [...map.values()];
+}
+
+function mergeGroupsBySlugs<T extends Group>(
+  a: T[] | undefined,
+  b: T[] | undefined,
+) {
+  const map = new Map<string, T>();
+
+  const addToMap = (groups: T[]) => {
```

Comment:
If possible move the function out. Types, and execution will be faster and the code is more pure.

Reply:
I don't think this function is easily extracted, the reason it's inlined is precisely because it isn't pure and mutates the `map` variable from outer scope via closure. Would require refactoring to `[a, b].reduce(...)` and use POJO instead of `Map` to make it truly immutable. But I don't think that's necessary, mutations are OK as long as they're restricted to local function implementation and function is pure from the outside.